applications/protocols_serialization/client/src/ot_shell.c

```diff
@@ -19,16 +23,244 @@ static int ot_console_cb(void *context, const char *format, va_list arg)
+{
+	if (argc <= 1) {
+		shell_print(sh, "%s", otIp6IsEnabled(NULL) ? "up" : "down");
+		return OT_ERROR_NONE;
+	}
+
+	if (strcmp(argv[1], "up") == 0) {
+		return otIp6SetEnabled(NULL, true);
+	}
+
+	if (strcmp(argv[1], "down") == 0) {
+		return otIp6SetEnabled(NULL, false);
+	}
+
+	return OT_ERROR_INVALID_COMMAND;
+}
+
+static otError ot_cli_command_ipmaddr(const struct shell *sh, size_t argc, char *argv[])
+{
+	struct in6_addr in6_addr;
+
+	if (argc <= 1) {
+		const otNetifMulticastAddress *addr;
+		char addr_string[NET_IPV6_ADDR_LEN];
+
+		for (addr = otIp6GetMulticastAddresses(NULL); addr != NULL; addr = addr->mNext) {
+			memcpy(in6_addr.s6_addr, addr->mAddress.mFields.m8, OT_IP6_ADDRESS_SIZE);
+
+			if (!net_addr_ntop(AF_INET6, &in6_addr, addr_string, sizeof(addr_string))) {
+				return OT_ERROR_FAILED;
+			}
+
+			shell_print(sh, "%s", addr_string);
+		}
+
+		return OT_ERROR_NONE;
+	}
+
+	if (argc == 3 && strcmp(argv[1], "add") == 0) {
+		otIp6Address addr;
+
+		if (net_addr_pton(AF_INET6, argv[2], &in6_addr)) {
+			return OT_ERROR_INVALID_ARGS;
+		}
+
+		memcpy(addr.mFields.m8, in6_addr.s6_addr, OT_IP6_ADDRESS_SIZE);
+
+		return otIp6SubscribeMulticastAddress(NULL, &addr);
+	}
+
+	if (argc == 3 && strcmp(argv[1], "del") == 0) {
+		otIp6Address addr;
+
+		if (net_addr_pton(AF_INET6, argv[2], &in6_addr)) {
+			return OT_ERROR_INVALID_ARGS;
+		}
+
+		memcpy(addr.mFields.m8, in6_addr.s6_addr, OT_IP6_ADDRESS_SIZE);
+
+		return otIp6UnsubscribeMulticastAddress(NULL, &addr);
+	}
+
+	return OT_ERROR_INVALID_ARGS;
+}
+
+static otError ot_cli_command_mode(const struct shell *sh, size_t argc, char *argv[])
+{
+	otLinkModeConfig mode;
+
+	if (argc <= 1) {
+		char mode_string[3];
+		size_t i = 0;
+
+		mode = otThreadGetLinkMode(NULL);
+
+		if (mode.mRxOnWhenIdle) {
+			mode_string[i++] = 'r';
+		}
+
+		if (mode.mDeviceType) {
+			mode_string[i++] = 'd';
+		}
+
+		if (mode.mNetworkData) {
+			mode_string[i++] = 'n';
+		}
+
+		if (i == 0) {
+			mode_string[i++] = '-';
+		}
+
+		shell_print(sh, "%.*s", i, mode_string);
+		return OT_ERROR_NONE;
+	}
+
+	memset(&mode, 0, sizeof(mode));
+
+	if (strcmp(argv[1], "-") != 0) {
+		for (const char *c = argv[1]; *c; c++) {
+			switch (*c) {
+			case 'r':
+				mode.mRxOnWhenIdle = true;
+				break;
+			case 'd':
+				mode.mDeviceType = true;
+				break;
+			case 'n':
+				mode.mNetworkData = true;
+				break;
+			default:
+				return OT_ERROR_INVALID_ARGS;
+			}
+		}
+	}
+
+	return otThreadSetLinkMode(NULL, mode);
+}
+
+static otError ot_cli_command_pollperiod(const struct shell *sh, size_t argc, char *argv[])
+{
+	if (argc <= 1) {
+		shell_print(sh, "%u", (unsigned int)otLinkGetPollPeriod(NULL));
+		return OT_ERROR_NONE;
+	}
+
+	return otLinkSetPollPeriod(NULL, strtoul(argv[1], NULL, 10));
+}
+
+static otError ot_cli_command_state(const struct shell *sh, size_t argc, char *argv[])
+{
+	static const char *const role_str[] = {"disabled", "detached", "child", "router", "leader"};
```

Comment:
why is it static?

Reply:
it makes sense that it be in `rodata` instead of being constructed on stack on each call, no?

Reply:
Names will be located in rodata even without static, the only array with pointers is constructed on stack (5 * 4) = 20 bytes. That doesn't seem a lot.

Reply:
Yes, I meant the array, of course. The assembly is a bit shorter that way. Is there a specific reason you don't like `static`?

Reply:
I do not like static if it is allocated on ram and static that consumes ram for rarely used function. If it is in flash then it is Ok.

Reply:
Ok, I agree but in this case it's in flash.